doc/read-the-docs-site/architecture/key-concepts.rst

```diff
@@ -1,6 +1,251 @@
+important bits:
+
+- We hide the indexer representation under the worker type. Any concrete indexer
+  type will work as long as it implements the ``IsIndex`` typeclass.
+- The indexer is put in an ``MVar`` to allow access to it from other threads.
+  We'll come back to it later.
+- We provide a preprocessor, which can be seen as a functon that transforms
+  the input send by the coordinator into events.
+- ``hoistError`` ensures that we know how to translate the base monad of the
+  indexer into ``ExceptT IndexerError IO`` which is the base monad of a worker.
+
+Once we have a list of workers, we can use it to create a coordinator using the
+``mkCoordinator`` function.
+This function inialise each worker, creating a dedicated thread for each worker
+where the worker will wait for incoming action to perform and notify the
+coordinator when the action is performed.
+
+The coordinator monitors each of the threads and, if one of the worker encounter
+an error, it will try to close all the other workers nicely.
+
+``Coordinator`` itself implements ``IsIndex`` and thus we can itself be wrapped
+in a worker.
+Thanks to it, we can create a whole hierarchy of indexers that can control from
+a main coordinator.
+
+
+Preprocessor
+------------
+
+We saw in the coordinator and workers section that workers take a preprocessor.
+As stated in this section, the preprocessor type can be viewed as a stateful
+function that transform the action sent to an indexer.
+It's type is isomorphic to:
+
+.. code-block:: haskell
+
+   StateT s m ([ProcessedInput point a] -> [ProcessedInput point b])
+
+It's the first time we encounter ``ProcessedInput`` so it is worth going through
+its definition:
+
+
+.. code-block:: haskell
+
+   data ProcessedInput point event
+     = Rollback point
+     | Index (Timed point (Maybe event))
+     | IndexAllDescending (NonEmpty (Timed point (Maybe event)))
+     | StableAt point
+     | Stop
+
+It is mostly a reified version (functions expressed as data) of most of the
+``IsIndex`` functions plus a ``Stop`` construct that allows us to stop a worker.
+So the goal of a preprocessor is to take a list of actions that must be sent to
+a worker and to transform this list.
+It can be either to filter out some actions or to add some actions to the list,
+based on the internal state ``s`` of the preprocessor.
+
+Transformers
+------------
+
+Indexer transformers (or, shorter, transformers) are used to alter the behaviour
+of an existing indexer.
+The name comes from the monad transformers concept, and it was chosen because
+monad and indexer transformers have similarities, as both aims at adding
+capabilities to a base implementation.
+
+A transformer carries a state and an underlying indexer.
+The typeclass instances of the transformer can then add extra logic to this
+underlying indexer.
+
+Preprocessor vs Transformers
+----------------------------
+
+There's a lot of similarities between preprocessor and transformers.
```

Comment:
Can some preprocessors be rewritten as transformers? Or vis-versa? Maybe mention when that would be the case.

Reply:
Yes, added.